### PR TITLE
Mrtk version next controller mapping

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
Overview
---
Adds the capability to create custom controller mappings as another configuration window.  Lends heavily (almost identical) to @StephenHodgson 's work.  

Also adds the empty definitions for several well known controller types to aid configuration profile testing (but do not take this as read that full support for each is forthcoming for Alpha)

The interfaces for Devices and Controllers has also been added to support this work (taken from #2184

Changes
---
- Adds Device / Controller Interfaces
- Re-implemented the Controller Mapping profile, including the option to configure a model per controller
- Added the Controller Mapping Inspector
- Updated the Main configuration inspector to add support for controller mapping and default controller rendering
- General tidyup
- Added Empty device definitions for common style controllers

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs [hunk 1]
```diff
@@ -756,6 +582,200 @@ public void EnableManager(Type type, string managerName)
             return (T)manager;
         }
 
+        #endregion Individual Manager Management
+
+        #region Multiple Managers Management
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: Idk why, but Managers Management made me chuckle a little.

What about `Multiple Manager Utilities`?

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: rofl, now we are strategizng on #if names :P  
Sure, I'll quickly change that so we can progress.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: 😋

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ protected bool CheckMixedRealityManager()
                 // Search the scene for one, in case we've just hot reloaded the assembly.
                 var managerSearch = FindObjectsOfType<MixedRealityManager>();
 
-                if (managerSearch.Length == 0)
+                if (managerSearch == null || managerSearch.Length == 0)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs]: `managerSearch` can never be null. I fixed this in my housekeeping branch. please revert.

File: Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs [hunk 3]
```diff
@@ -92,7 +92,20 @@ public void ResetConfiguration(MixedRealityConfigurationProfile profile)
...
+        #endregion Mixed Reality runtime component registry
+
+        #region Active SDK components
+
+        /// <summary>
+        /// The Active Device property maintains the SDK detected by the Mixed Reality manager on startup
+        /// </summary>
+        [SerializeField]
+        [Tooltip("The currently active / detected Headset or SDK")]
+        private IMixedRealityDevice activeDevice = default(IMixedRealityDevice);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: Is it possible to have more than one active at a time?

File: Assets/MixedRealityToolkit-SDK/Models/Cube.prefab [hunk 4]
```diff
@@ -0,0 +1,98 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Models/Cube.prefab]: can we do something other than a cube?

File: Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs [hunk 5]
```diff
@@ -103,6 +104,29 @@ public MixedRealitySpeechCommandsProfile SpeechCommandsProfile
...
+
+        [SerializeField]
+        [Tooltip("Device profile for wiring up physical inputs to Actions.")]
+        private MixedRealityControllerMappingProfile controllersProfile;
+        public MixedRealityControllerMappingProfile ControllersProfile
+        {
+            get { return controllersProfile; }
+            private set { controllersProfile = value; }
+        }
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs]: extra line space

File: Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs [hunk 6]
```diff
@@ -103,6 +104,29 @@ public MixedRealitySpeechCommandsProfile SpeechCommandsProfile
...
+        public bool EnableControllerProfiles
+        {
+            get { return enableControllerProfiles; }
+            private set { enableControllerProfiles = value; }
+        }
+
+        [SerializeField]
+        [Tooltip("Device profile for wiring up physical inputs to Actions.")]
+        private MixedRealityControllerMappingProfile controllersProfile;
+        public MixedRealityControllerMappingProfile ControllersProfile
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs]: missing summary

File: Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs [hunk 7]
```diff
@@ -116,6 +140,32 @@ public bool RenderMotionControllers
...
         }
 
+        [SerializeField]
+        [Tooltip("Default Left Controller Model")]
+        private GameObject defaultLeftHandModel;
+
+        /// <summary>
+        /// The Default controller model when there is no specific controller model for the Left hand or when no hand is specified (Handedness = none)
+        /// </summary>
+        public GameObject DefaultLeftHandModel
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs]: Can we move this into the device template inspector?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMapping.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        /// </summary>
+        public InteractionMapping[] Interactions => interactions;
+
+        [SerializeField]
+        [Tooltip("Details the list of available buttons / interactions available from the device.")]
+        private InteractionMapping[] interactions;
+
+        public InteractionMapping GetInteractionMapping(DeviceInputType inputType)
+        {
+            return default(InteractionMapping);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMapping.cs]: What's this do?

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs (hunk 2 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs]: Already removed from your fix changes

File: Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs (hunk 3 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: This is scaffold only as we haven't got the SDK detection stuff in there yet
In reality, NO.  As you can only have one headset on at a time :D (unless Zaphod Beeblebrox wants to try it)

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityBaseConfigurationProfileInspector.cs]: odd, still showing up as a change for me.

File: Assets/MixedRealityToolkit-SDK/Models/Cube.prefab (hunk 4 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit-SDK/Models/Cube.prefab]: We could but MVP.
I'll see about using a different controller model in the devices PR.

File: Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs (hunk 5 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs]: Removed

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMapping.cs (hunk 8 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMapping.cs]: I'll remove it.  I'm sure there was a good reason originally but I forget now.

File: Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs (hunk 6 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/MixedRealityConfigurationProfile.cs]: Grr, that was your property, but my bad for not documenting it when I grabbed it

File: Assets/MixedRealityToolkit/_Core/Devices/Simulator/SimulatedController.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Devices;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Utilities;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Devices.Simulator
+{
+    public struct SimulatedController : IMixedRealityController
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Devices/Simulator/SimulatedController.cs]: Could we also add a simulated Hand controller as well?

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityControllerMappingProfileInspector.cs [hunk 10]
```diff
@@ -26,13 +26,13 @@ public class MixedRealityControllerMappingProfileInspector : MixedRealityBaseCon
...
         private static int[] actionIds;
 
+        private SerializedProperty renderMotionControllers;
+        private SerializedProperty useDefaultModels;
+        private SerializedProperty overrideLeftHandModel;
+        private SerializedProperty overrideRightHandModel;
+
         private void OnEnable()
         {
-            if (!CheckMixedRealityManager())
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityControllerMappingProfileInspector.cs]: we should probably keep this check here seeing as we're asking for the manager a few lines later.

File: Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityInputActionsProfileInspector.cs [hunk 11]
```diff
@@ -38,7 +38,10 @@ public override void OnInspectorGUI()
 
             EditorGUILayout.LabelField("Input Actions", EditorStyles.boldLabel);
 
-            CheckMixedRealityManager();
+            if (!CheckMixedRealityManager())
+            {
+                CheckMixedRealityManager();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/MixedRealityInputActionsProfileInspector.cs]: should be a return.

File: Assets/MixedRealityToolkit/_Core/Devices/Simulator/SimulatedController.cs (hunk 9 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/Simulator/SimulatedController.cs]: Done

**Pull request merged**